iml-gui/crate/src/dependency_tree.rs

```diff
@@ -0,0 +1,444 @@
+///
+/// Please note that the contents of `deps` and `dset` should be always equal, i.e.
+/// ```norun
+/// let set1 = HashSet::from_iter(self.deps.iter().cloned();
+/// let set2 = self.deps;
+/// asssert_eq!()
+/// ```
+#[derive(Clone, Debug)]
+pub struct Rich<K: Hash + Eq, T> {
+    pub id: K,
+    pub deps: Vec<K>,
+    pub dset: HashSet<K>,
+    pub inner: T,
+}
+
+impl<K, T> Rich<K, T>
+where
+    K: Hash + Ord + Copy,
+    T: Clone,
+{
+    pub fn new(t: T, extract: impl FnOnce(&T) -> (K, Vec<K>)) -> Self {
+        let (id, mut deps) = extract(&t);
+        deps.sort();
+        Self {
+            id,
+            dset: HashSet::from_iter(deps.iter().cloned()),
+            deps,
+            inner: t,
+        }
+    }
+}
+
+impl<K, T> Deps<K> for Rich<K, T>
+where
+    K: Hash + Ord + Copy,
+{
+    fn id(&self) -> K {
+        self.id
+    }
+    fn deps(&self) -> &[K] {
+        &self.deps
+    }
+    fn has(&self, k: &K) -> bool {
+        self.dset.contains(k)
+    }
+}
+
+impl<K, T> Deref for Rich<K, T>
+where
+    K: Hash + Ord + Copy,
+{
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct DependencyDAG<K: Hash + Eq + Debug, T> {
+    pub roots: Vec<Arc<T>>,
+    pub deps: HashMap<K, Vec<Arc<T>>>,
+}
+
+impl<K: Hash + Eq + Debug, T> Default for DependencyDAG<K, T> {
+    fn default() -> Self {
+        Self {
+            roots: Vec::new(),
+            deps: HashMap::new(),
+        }
+    }
+}
+
+impl<K: Hash + Eq + Debug, T> DependencyDAG<K, T> {
+    pub fn clear(&mut self) {
+        self.roots.clear();
+        self.deps.clear();
+    }
+}
+
+pub fn build_direct_dag<K, T>(ts: &[Arc<T>]) -> DependencyDAG<K, T>
+where
+    K: Hash + Eq + Copy + Debug,
+    T: Deps<K> + Clone + Debug,
+{
+    let mut roots: Vec<K> = Vec::new();
+    let mut dag: Vec<(K, Vec<K>)> = Vec::with_capacity(ts.len());
+    if ts.len() == 1 {
+        // special case, when there is no any arc `x -> y`,
+        // then the only vertex becomes the root
+        roots.push(ts[0].id());
+    } else {
+        for t in ts {
+            let x = t.id();
+            let ys = t.deps();
+            if !ys.is_empty() {
+                // push the arcs `x -> y` for all `y \in xs` into the graph
+                dag.push((x, ys.to_vec()));
+                if !roots.contains(&x) {
+                    roots.push(x);
+                }
+                // remove any of the destinations from the roots
+                for y in ys {
+                    if let Some(i) = roots.iter().position(|r| *r == *y) {
+                        roots.remove(i);
+                    }
+                }
+            }
+        }
+    }
+    enrich_dag(ts, &roots, &dag)
+}
+
+pub fn build_inverse_dag<K, T>(ts: &[Arc<T>]) -> DependencyDAG<K, T>
+where
+    K: Hash + Eq + Copy + Debug,
+    T: Deps<K> + Clone + Debug,
+{
+    let mut roots: HashSet<K> = ts.iter().map(|t| t.id()).collect();
```

Comment:
Why does `build_inverse_dag` look simpler than `build_direct_dag`?

Reply:
Could you define _simpler_? They are almost the same in terms of code size and complexity, no?